Permit restoring mysql/postgresql DBs

rbtcollins opened the pull request:
Fixes: #209

The duplication in the creation of the Azure API object is undesirable
but I can't see any way around it short of serialising back to JSON and
up again via the Azure API JSON decode path, which seems more fragile
than the typed code path this provides.

Signed-off-by: Robert Collins <robert.collins@cognite.com>

### Description of your changes

Fixes #209 

I have:

- [X ] Read and followed Crossplane's [contribution process].
- [X ] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

I haven't yet testing this against a live Azure cluster - as with the fix for 210, I'd love some guidance on how to connect a dev build directly to a cluster - if someone can step me through it on slack I can write it up as a doc fix.

negz commented on apis/database/v1beta1/sql_types.go:
```diff
@@ -179,6 +184,18 @@ type SQLServerParameters struct {
 }
 
+// CreateMode controls the creation behaviour
+// +kubebuilder:validation:Enum=Default;GeoRestore;PointInTimeRestore;Replica;ServerPropertiesForCreate
```
Can you use kubebuilder validation comments on a type declaration like this? I've only ever seen them used on fields.

rbtcollins replied:
Yep; see https://book.kubebuilder.io/cronjob-tutorial/api-design.html for instance. And of course the generated code

negz commented on apis/database/v1beta1/sql_types.go:
```diff
@@ -18,6 +18,7 @@ package v1beta1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Nit: We import this as `metav1` by convention - see the above line.

rbtcollins replied:
will fix

negz reviewed (commented):
This looks good so far @rbtcollins! I commented on #210 with some details on how to manually test the changes. Can you also add some unit tests to cover the new create modes?

rbtcollins commented:
Righto, I've refactored the code to permit unit testing of the new case statement, fixed a nil pointer bug that was latent, added exhaustive unit tests for the case statement (though can I note that automatic fuzz testing would be a great time investment for this team to get comprehensive coverage with low developer investment...)

Working on actually hooking it up in a live cluster next

rbtcollins commented:
We hooked this up to azure and it sortof worked: it accepted a restore request but sat in 'request has been accepted' forever. I suspect we don't know how to use the SDK properly to restore yet.

```
2021-01-21T16:16:57.035+0100    DEBUG   provider-azure  Reconciling     {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01"}
2021-01-21T16:16:58.544+0100    DEBUG   provider-azure  Successfully requested creation of external resource    {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01", "uid": "e99671ab-f3be-47bb-86d1-ecb827139e20", "version": "370064226", "external-name": "assets-v3cluster01", "requeue-after": "2021-01-21T16:17:28.544+0100"}
2021-01-21T16:16:58.544+0100    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"PostgreSQLServer","name":"assets-v3cluster01","uid":"e99671ab-f3be-47bb-86d1-ecb827139e20","apiVersion":"database.azure.crossplane.io/v1beta1","resourceVersion":"370064226"}, "reason": "CreatedExternalResource", "message": "Successfully requested creation of external resource"}
2021-01-21T16:17:01.714+0100    DEBUG   provider-azure  Reconciling     {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01"}
2021-01-21T16:17:04.202+0100    DEBUG   provider-azure  Successfully requested creation of external resource    {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01", "uid": "e99671ab-f3be-47bb-86d1-ecb827139e20", "version": "370064448", "external-name": "assets-v3cluster01", "requeue-after": "2021-01-21T16:17:34.202+0100"}
2021-01-21T16:17:04.202+0100    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"PostgreSQLServer","name":"assets-v3cluster01","uid":"e99671ab-f3be-47bb-86d1-ecb827139e20","apiVersion":"database.azure.crossplane.io/v1beta1","resourceVersion":"370064448"}, "reason": "CreatedExternalResource", "message": "Successfully requested creation of external resource"}
2021-01-21T16:17:07.777+0100    DEBUG   provider-azure  Reconciling     {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01"}
2021-01-21T16:17:09.249+0100    DEBUG   provider-azure  Successfully requested creation of external resource    {"controller": "managed/postgresqlserver.database.azure.crossplane.io", "request": "/assets-v3cluster01", "uid": "e99671ab-f3be-47bb-86d1-ecb827139e20", "version": "370064549", "external-name": "assets-v3cluster01", "requeue-after": "2021-01-21T16:17:39.249+0100"}
2021-01-21T16:17:09.250+0100    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"PostgreSQLServer","name":"assets-v3cluster01","uid":"e99671ab-f3be-47bb-86d1-ecb827139e20","apiVersion":"database.azure.crossplane.io/v1beta1","resourceVersion":"370064549"}, "reason": "CreatedExternalResource", "message": "Successfully requested creation of external resource"}
```

rbtcollins commented:
We did finally get an error in the activity monitor:
```
 "properties": {
        "statusMessage": "{\"status\":\"Failed\",\"error\":{\"code\":\"ResourceOperationFailure\",\"message\":\"The resource operation completed with terminal provisioning state 'Failed'.\",\"details\":[{\"code\":\"InvalidConfigurationValue\",\"message\":\"The value '1/21/2021 2:00:14 PM' for configuration 'PointInTime' is not valid. The allowed values are '> 1/21/2021 2:20:56 PM and < 1/21/2021 3:17:26 PM'.\"}]}}",
        "eventCategory": "Administrative",
        "entity": "/subscriptions/531564ca-ed4c-48b5-b4d0-6363b20c3aeb/resourcegroups/assets-v2cluster01/providers/Microsoft.DBforPostgreSQL/servers/assets-v3cluster01",
        "message": "Microsoft.DBforPostgreSQL/servers/write",
        "hierarchy": "531564ca-ed4c-48b5-b4d0-6363b20c3aeb"
    },
    ```
    
    So I think this would have worked with different inputs

negz commented on apis/database/v1beta1/sql_types.go:
```diff
@@ -162,7 +162,14 @@ type SQLServerParameters struct {
+	RestorePointInTime *metav1.Time `json:"restorePointInTime,omitempty"`
+
+	// SourceServerID - The server to restore from when restoring or creating replicas
+	SourceServerID *string `json:"sourceServerID,omitempty"`
```
Another optional for `CreateMode` might be to make it a required field, but have it default (using the `// +kubebuilder:default` comment marker) to `CreateModeDefault`.

rbtcollins replied:
We tested the minimal Tls behaviour with this patch applied as well, from a unified branch, so we know default creates still work.

The zero value for string is "" rather than nil, which just makes handling of the empty case segfault-safe - a nice property to have - and this also lines up with the Azure SDK better. Of course, it is your code base and I can change it, but it will be more code, less ergonomic on the go side, and won't change the Kubernetes API side one iota.

Re: optional - thats only needed when fields have been marked Required, and nothing in this package has been tagged as required to kubebuilder - all the fields are currently optional as far as I can tell - and I've manually verified that this is the case by testing with and without these fields.

Probably we want to fix that, but I think we should do that in one hit, rather than having a mix of correctly annotated and not-annotated fields, because that will be harder to reason about.

negz replied:
> We tested the minimal Tls behaviour with this patch applied as well, from a unified branch, so we know default creates still work.

Nice - thank you.

> Re: optional - thats only needed when fields have been marked Required, and nothing in this package has been tagged as required to kubebuilder - all the fields are currently optional as far as I can tell - and I've manually verified that this is the case by testing with and without these fields.

Are we both talking about the API conventions around [optional vs required fields](https://github.com/kubernetes/community/blob/81ece459498964d18159d00948bd4edb849bc8c1/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required)? These state that all optional fields should be pointer types with the `omitempty` JSON struct tag and the `// +optional` comment marker. As far as I can tell this file currently follows those conventions. In this case I believe controller-tools is able to infer that the fields are optional from the `omitempty` tags alone, but I'd like to include the comment markers too for consistency.

rbtcollins replied:
We weren't. 

On +optional : the rest of the file definitely doesn't follow that consistently, though I can see its present in some places. I'll add it to this and ping when I've done so.

On pointer: I understand the theory of telling 'not supplied string' vs 'empty string' apart, but the Azure SDK we're using doesn't distinguish - an unsupplied string (nil) needs to be converted to an empty string by ForProvider. And in `IsPostgreSQLUpToDate` similarly we'll need to compare a 'nil or ""' string on the Crossplane side to a "" string on the Azure side, always - so I'd really rather not do that, since we cannot use that distinction anywhere. Is that ok?

negz replied:
> And in IsPostgreSQLUpToDate similarly we'll need to compare a 'nil or ""' string on the Crossplane side to a "" string on the Azure side, always - so I'd really rather not do that, since we cannot use that distinction anywhere. Is that ok?

No, sorry. I understand where you're coming from with regard to it being technically fairly pointless but I'd like us to comply with the API conventions of the ecosystem we're living in. Having as much consistency as possible across our providers helps us when we get into things generating these types automatically.

negz reviewed (commented):
Got it! Can you validate that default creates still work as expected? Other than a rebase and the one comment I think that's all we need to get this merged.

rbtcollins commented:
@negz ping ?

rbtcollins commented on pkg/clients/database/mysql.go:
```diff
@@ -82,30 +83,95 @@ func (c *MySQLServerClient) GetServer(ctx context.Context, cr *azuredbv1beta1.My
+// toMySQLProperties converts the CrossPlane ForProvider object to a MySQL Azure properties object
+func toMySQLProperties(s v1beta1.SQLServerParameters, adminPassword string) mysql.BasicServerPropertiesForCreate {
+	switch s.CreateMode {
+	case azuredbv1beta1.CreateModeDefault:
```
fallthrough is more elegant, done.

rbtcollins commented:
@negz all updated if you'd care to re-review

negz commented:
@rbtcollins Did you push a new update since the last review? I'm not seeing any changes.

rbtcollins commented:
Yes, github has lost its mind. https://github.com/cognitedata/provider-azure/commit/64518728c0fc6b327d0748db3da02bdecbc185cf is the most recent commit in our branch https://github.com/cognitedata/provider-azure/tree/restore-db and I have no idea why the PR isn't showing the update.

rbtcollins commented:
ok, and that autoclose-due-to-merge-detected was annoying (I pushed HEAD^ as part of trying to trigger a change detection in github)

rbtcollins commented:
There we go, eventually consistent hah.  Hit it with a stick and it will be.

negz reviewed (approved):
Thanks @rbtcollins! I bet this is going to be useful to a lot of people.
